File: binding/ios/CheetahManager.swift
//
//  Copyright 2022 Picovoice Inc.
//  You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
//  file accompanying this source.
//  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
//  specific language governing permissions and limitations under the License.
//

import ios_voice_processor

/// High-level iOS binding for Cheetah speech-to-text engine. It handles recording audio from microphone, processes it
/// in real-time using Cheetah, and notifies the client with transcriptions made from audio.
public class CheetahManager {
    private var onTranscription: ((String) -> Void)?
    private var errorCallback: ((CheetahError) -> Void)?

    private var cheetah: Cheetah?
    private var isListening = false

    /// Private constructor.
    private init(cheetah: Cheetah, onTranscription: ((String) -> Void)?, errorCallback: ((CheetahError) -> Void)?) throws {
        self.onTranscription = onTranscription
        self.errorCallback = errorCallback
        self.cheetah = cheetah
    }

    /// Constructor.
    ///
    /// - Parameters:
    ///   - accessKey: The AccessKey obtained from Picovoice console.
    ///   - modelPath: Absolute path to file containing model parameters.
    ///   - endpointDuration: Duration of endpoint in seconds. A speech endpoint is detected when there is a
    ///         chunk of audio (with a duration specified herein) after an utterance without any speech in it. Default
    ///         is 1 second.
    ///   - onTranscription: It is invoked if a non empty transcription is return from Cheetah.
    ///   - errorCallback: Invoked if an error occurs while processing frames. If missing, error will be printed to console.
    /// - Throws: CheetahError
    public convenience init(
        accessKey: String,
        modelPath: String? = nil,
        endpointDuration: Flaot = 1.0,
        onTranscription: ((Int32) -> Void)?,
        errorCallback: ((Error) -> Void)? = nil) throws {

        try self.init(
                cheetah: Cheetah(accessKey: accessKey, modelPath: modelPath, endpointDuration: endpointDuration),
                onTranscription: onTranscription,
                errorCallback: errorCallback)
    }

    deinit {
        self.delete()
    }

    /// Stops recording and releases Cheetah resources
    public func delete() {
        if isListening {
            stop()
        }

        if self.cheetah != nil {
            self.cheetah!.delete()
            self.cheetah = nil
        }
    }

    ///  Starts recording audio from the microphone and processes frames through Cheetah.
    ///
    /// - Throws: CheetahError if microphone permission is not granted or Cheetah has been disposed.
    public func start() throws {

        guard !isListening else {
            return
        }

        if cheetah == nil {
            throw CheetahInvalidStateError("Cheetah must be initialized before processing")
        }

        // Only check if it's denied, permission will be automatically asked.
        guard try VoiceProcessor.shared.hasPermissions() else {
            throw CheetahRuntimeError("Microphone permissions denied")
        }

        try VoiceProcessor.shared.start(
                frameLength: Cheetah.frameLength,
                sampleRate: Cheetah.sampleRate,
                audioCallback: self.audioCallback
        )

        isListening = true
    }

    /// Stops recording from microphone
    public func stop() {
        guard isListening else {
            return
        }

        VoiceProcessor.shared.stop()

        isListening = false
    }

    /// Callback to run after after voice processor processes frames.
    private func audioCallback(pcm: [Int16]) {
        guard self.cheetah != nil else {
            return
        }

        do{
            var (transcription, endpoint) = try self.cheetah!.process(pcm: pcm)
            if endpoint {
                transcription += try self.cheetah!.flush()
            }
            if transcription.count > 0 {
                DispatchQueue.main.async {
                    self.onTranscription?(transcription)
                }
            }
        } catch {
            if self.errorCallback != nil {
                self.errorCallback!(error)
            } else {
                print("\(error)")
            }
        }
    }
}

File: demo/flutter/lib/cheetah_manager.dart
//
// Copyright 2022-2023 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
// file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//

import 'dart:async';
import 'package:cheetah_flutter/cheetah.dart';
import 'package:cheetah_flutter/cheetah_error.dart';
import 'package:flutter_voice_processor/flutter_voice_processor.dart';
import 'package:flutter/services.dart';

typedef TranscriptCallback = Function(String transcript);

typedef ProcessErrorCallback = Function(CheetahException error);

class CheetahManager {
  VoiceProcessor? _voiceProcessor;
  Cheetah? _cheetah;

  final TranscriptCallback _transcriptCallback;

  static Future<CheetahManager> create(
      String accessKey,
      String modelPath,
      TranscriptCallback transcriptCallback,
      ProcessErrorCallback processErrorCallback) async {
    Cheetah cheetah = await Cheetah.create(accessKey, modelPath,
        enableAutomaticPunctuation: true);
    return CheetahManager._(cheetah, transcriptCallback, processErrorCallback);
  }

  CheetahManager._(this._cheetah, this._transcriptCallback,
      ProcessErrorCallback processErrorCallback)
      : _voiceProcessor = VoiceProcessor.instance {
    _voiceProcessor?.addFrameListener((List<int> frame) async {
      if (!(await _voiceProcessor?.isRecording() ?? false)) {
        return;
      }
      if (_cheetah == null) {
        processErrorCallback(CheetahInvalidStateException(
            "Cannot process with Cheetah - resources have already been released"));
        return;
      }

      try {
        CheetahTranscript partialResult = await _cheetah!.process(frame);

        if (partialResult.isEndpoint) {
          CheetahTranscript remainingResult = await _cheetah!.flush();
          String finalTranscript =
              partialResult.transcript + remainingResult.transcript;
          if (remainingResult.transcript.isNotEmpty) {
            finalTranscript += " ";
          }
          _transcriptCallback(finalTranscript);
        } else {
          _transcriptCallback(partialResult.transcript);
        }
      } on CheetahException catch (error) {
        processErrorCallback(error);
      }
    });

    _voiceProcessor?.addErrorListener((VoiceProcessorException error) {
      processErrorCallback(CheetahException(error.message));
    });
  }

  Future<void> startProcess() async {
    if (await _voiceProcessor?.isRecording() ?? false) {
      return;
    }

    if (_cheetah == null || _voiceProcessor == null) {
      throw CheetahInvalidStateException(
          "Cannot start Cheetah - resources have already been released");
    }

    if (await _voiceProcessor?.hasRecordAudioPermission() ?? false) {
      try {
        await _voiceProcessor?.start(
            _cheetah!.frameLength, _cheetah!.sampleRate);
      } on PlatformException catch (e) {
        throw CheetahRuntimeException(
            "Failed to start audio recording: ${e.message}");
      }
    } else {
      throw CheetahRuntimeException(
          "User did not give permission to record audio.");
    }
  }

  Future<void> stopProcess() async {
    if (_cheetah == null || _voiceProcessor == null) {
      throw CheetahInvalidStateException(
          "Cannot start Cheetah - resources have already been released");
    }

    if (await _voiceProcessor?.isRecording() ?? false) {
      try {
        await _voiceProcessor?.stop();
      } on PlatformException catch (e) {
        throw CheetahRuntimeException(
            "Failed to stop audio recording: ${e.message}");
      }

      CheetahTranscript cheetahTranscript = await _cheetah!.flush();
      _transcriptCallback((cheetahTranscript.transcript) + " ");
    }
  }

  Future<void> delete() async {
    await stopProcess();
    _voiceProcessor = null;

    _cheetah?.delete();
    _cheetah = null;
  }
}
